Terrain types

GeorgianaIon opened the pull request:
Lava functionality, improved old icy functionality

GeorgianaIon committed:
improved icy functionality

GeorgianaIon committed:
solved motion bug with icy terrain

GeorgianaIon committed:
conflicts

GeorgianaIon committed:
merged main into my work

GeorgianaIon committed:
Lava terrain with fire

I moved the terrain(icy and lava) functionality in another class

GeorgianaIon committed:
Added health

GeorgianaIon committed:
Merge branch 'main' into terrain-types

GeorgianaIon committed:
Updates

Terrain updates, the physics material is no longer on the cat, it is on the terrain

GeorgianaIon committed:
Update terrainScene.unity

RealFoXsTeR committed:
Testing different tile properties

GeorgianaIon committed:
Fixed bug with the cat getting stuck

GeorgianaIon committed:
Merge remote-tracking branch 'origin/terrain-types-test' into terrain-types

GeorgianaIon committed:
Changed the tiles texture to smaller tiles

GeorgianaIon committed:
Made the terrains prefabs

GeorgianaIon committed:
Refactored Lava functionality

GeorgianaIon committed:
Merge branch 'main' into terrain-types

GeorgianaIon committed:
Solved Conflicts

RealFoXsTeR commented on Assets/Prefabs/TerrainPrefabs/Terrain.prefab:
```diff
@@ -0,0 +1,226 @@
+%YAML 1.1
```
![image](https://user-images.githubusercontent.com/72012358/230225978-3aaa15cf-d2bc-4162-8800-a3c4c3d4e23d.png)
Be aware of this warning. I'm also not sure if a tilemap really should be a prefab in the first place anyway

RealFoXsTeR commented on Assets/Scripts/Tiles/IceTile.cs:
```diff
@@ -0,0 +1,23 @@
+using UnityEngine;
+
+public class IceTile : MonoBehaviour, ITile
+{
+  public void OnEnter(PlayerController collision)
+  {
+    Rigidbody2D playerBody = collision.GetComponent<Rigidbody2D>();
+    collision.isOnIce = true;
+    playerBody.drag = 0f;
+  }
+
+  public void OnExit(PlayerController collision)
+  {
+    Rigidbody2D playerBody = collision.GetComponent<Rigidbody2D>();
+    collision.isOnIce = false;
+    playerBody.drag = 4f;
+  }
+  
+  public void onStay(PlayerController collision)
+  {
+    // Do nothing
+  }
+}
```
Probably want to split up the ITile interface into 2 so you don't violate the I in the SOLID principles.
Could also argue to name the variable controller or something rather than collision to better represent what the object is.

RealFoXsTeR commented on Assets/Scripts/Tiles/LavaTile.cs:
```diff
@@ -0,0 +1,39 @@
+
+public class LavaTile : MonoBehaviour, ITile
+{
+  [SerializeField]
+  private GameObject fireParticleSystem;
+
+  private ParticleSystem particleSystem;
```
The duplicate particle system should not be necessary.
Also note the little info message you get:
![image](https://user-images.githubusercontent.com/72012358/230228847-bc5bce85-dea4-4f4c-b6ca-11677bb0abef.png)

RealFoXsTeR commented on Assets/Scripts/Tiles/LavaTile.cs:
```diff
@@ -0,0 +1,39 @@
+  public void OnEnter(PlayerController collision)
+  {
+    var health = collision.gameObject.GetComponent<Health>();
+    // GameObject fireParticleSystem = collision.fireParticleSystem;
```
Just remove the outcommented code, unless you are going to change it/use it in the near future :)
Apparently there's also a "TryGetComponent" method that can be used instead if you are doing a null check afterwards anyway. Same applies for the 2 occurences in the PlayerController class.

RealFoXsTeR commented on Assets/Scripts/Tiles/LavaTile.cs:
```diff
@@ -0,0 +1,39 @@
+  {
+    var health = collision.gameObject.GetComponent<Health>();
+    // GameObject fireParticleSystem = collision.fireParticleSystem;
+    if (health != null)
+    {
+      health.TakeDamage(1);
+    }
+    particleSystem.Play();
```
Keep in mind, the player will currently only take damage upon entering the collider. Thus, you can stay in the lava for an unlimited amount of time without taking any further damage. Would probably make sense to take damage every X seconds or something. You might be able to use [InvokeRepeating](https://docs.unity3d.com/ScriptReference/MonoBehaviour.InvokeRepeating.html) on OnEnter and [CancelInvoke](https://docs.unity3d.com/ScriptReference/MonoBehaviour.CancelInvoke.html) on OnExit. An alternative could be coroutines with a boolean flag to check if the player is "onLava" or not.

RealFoXsTeR commented on Assets/Scripts/Player/PlayerController.cs:
```diff
@@ -47,12 +49,13 @@ private void FixedUpdate()
-    if (isSlipperyMovement && IsPlayerGrounded())
+    if (isOnIce && IsPlayerGrounded())
     {
+      // float horizontalDirection = Input.GetAxis("Horizontal");
+      // body.AddForce(new Vector2(horizontalDirection * moveSpeed, 0f), ForceMode2D.Force);
```
Again, just delete this.

RealFoXsTeR commented on Assets/Scripts/Player/PlayerController.cs:
```diff
@@ -3,8 +3,10 @@
+  //[Header("General settings")]
   private Rigidbody2D body;
-  private SpriteRenderer spriteRenderer;
+  public bool isOnIce = false;
```
No public bool. If it should be accessible by the IceTile class, make it a property instead. Could also argue to remove this and use the "isSlipperyMovement" instead as that is more generic and could be applied to a "SlimeTile" or something to make that slippery too.

GeorgianaIon committed:
Improvements

splited ITile interface into two to not violate the I in SOLID; now the cat takes damage avery 2 seconds spent on lava, Removed the unnecessary use of the GameObject of the fire particle system from LavaTile

GeorgianaIon committed:
terrain prefab update transform

GeorgianaIon committed:
Merge branch 'main' into terrain-types

GeorgianaIon committed:
Solved conflict

RealFoXsTeR commented on Assets/Scripts/Tiles/LavaTile.cs:
```diff
@@ -0,0 +1,57 @@
+  public void OnEnter(PlayerController playerController)
+  {
+    fireParticleSystem.Play();
+    TakeBurnDamage(playerController);
```
Remove this call? Will (in theory - not actually in practice as a mechanism preventing this from happening was implemented in Health.cs as well) damage player 2 times on entering the tile as coroutine is started immediately after which also calls TakeBurnDamage.

RealFoXsTeRDOTA commented on Assets/Scripts/Tiles/LavaTile.cs:
```diff
@@ -0,0 +1,57 @@
+    
+  }
+
+  public void onStay(PlayerController playerController)
```
If I was gonna be really nit-picky, you should use PascalCase for method names and also remove whitespace as in line 9 or 32.

GeorgianaIon committed:
Update LavaTile.cs

Removed the necessary TakeDamage call
